core/file_api.php

```diff
@@ -98,45 +98,49 @@ function file_get_display_name( $p_filename ) {
 }
 
 /**
- * Check the number of attachments a bug has (if any)
- * @param integer $p_bug_id A bug identifier.
- * @return integer
+ * Fills the cache with the attachement count from a list of bugs
+ * using one sql query
+ * @param array $p_bug_ids Array of bug ids
  */
-function file_bug_attachment_count( $p_bug_id ) {
+function file_bug_attachment_count_cache( array $p_bug_ids ){
 	global $g_cache_file_count;
 
-	# First check if we have a cache hit
-	if( isset( $g_cache_file_count[$p_bug_id] ) ) {
-		return $g_cache_file_count[$p_bug_id];
+	if( null === $p_bug_ids || !is_array( $p_bug_ids ) ){
+		return;
 	}
 
-	# If there is no cache hit, check if there is anything in
-	#   the cache. If the cache isn't empty and we didn't have
-	#   a hit, then there are not attachments for this bug.
-	if( count( $g_cache_file_count ) > 0 ) {
-		return 0;
-	}
+	$t_sql_list= '(' . implode( ', ', $p_bug_ids ) . ')';
+	$t_query = 'SELECT B.id AS bug_id, COUNT(F.bug_id) AS attachments'
+			. ' FROM {bug} B LEFT OUTER JOIN {bug_file} F'
+			. ' ON (B.id = F.bug_id)'
+			. ' WHERE B.id IN ' . $t_sql_list . ' GROUP BY B.id';
 
```

Comment:
> Instinctively I'd go for temp tables myself.

defining an api funcion, use a persistent table? (create is not an enforced permission in mantis db user), allow concurrency, etc 

> IMO partitioning the data would be resource-intensive and complex to develop / maintain.

Unless the method for insertion itself has to partition the data too. When Mysql fails its probably due to message size, not a limit for IN elements. Would it fail too if the insert is also a single "bulk" insert?

This would be the best approach, but you need to open a RFC for developers to discuss.

In the mean time, a mid-term solution, seems to be only partitioning the query for each case?